base/shell/explorer/traywnd.cpp

```diff
@@ -264,6 +264,199 @@ class CStartButton
 
 };
 
+// This window class name is CONFIRMED on Win10 by WinHier.
+static const WCHAR szTrayShowDesktopButton[] = L"TrayShowDesktopButtonWClass";
+
+// The 'Show Desktop' button at edge of taskbar
+class CTrayShowDesktopButton :
+    public CWindowImpl<CTrayShowDesktopButton, CWindow, CControlWinTraits>
+{
+    LONG m_nClickedTime;
+    BOOL m_bHovering;
+
+public:
+    DECLARE_WND_CLASS_EX(szTrayShowDesktopButton, CS_HREDRAW | CS_VREDRAW, COLOR_3DFACE)
+
+    CTrayShowDesktopButton() : m_nClickedTime(0), m_bHovering(FALSE)
+    {
+    }
+
+    INT WidthOrHeight() const
+    {
+#define SHOW_DESKTOP_MINIMUM_WIDTH 6
+        INT cxy = 2 * ::GetSystemMetrics(SM_CXEDGE);
+        return ((cxy >= SHOW_DESKTOP_MINIMUM_WIDTH) ? cxy : SHOW_DESKTOP_MINIMUM_WIDTH);
+#undef SHOW_DESKTOP_MINIMUM_WIDTH
+    }
+
+    HRESULT DoCreate(HWND hwndParent, BOOL bShow)
+    {
+        DWORD style = WS_CHILD | WS_CLIPSIBLINGS;
+        if (bShow)
+            style |= WS_VISIBLE;
+
+        Create(hwndParent, NULL, NULL, style);
+        if (!m_hWnd)
+            return E_FAIL;
+
+        if (::SetWindowTheme(m_hWnd, L"Start", NULL) != S_OK)
+            ::SetWindowTheme(m_hWnd, L"Explorer", NULL);
+        return S_OK;
+    }
+
+    LRESULT OnClick(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+    {
+        // The actual action can be delayed as an expected behaviour.
+        // But a too late action is an unexpected behaviour.
+        LONG nTime0 = m_nClickedTime;
+        LONG nTime1 = ::GetMessageTime();
+        if (nTime1 - nTime0 >= 600) // Ignore after 0.6 sec
+            return 0;
+
+        INPUT inputs[4];
+        ZeroMemory(inputs, sizeof(inputs));
+
+        // Emulate Win+D
```

Comment:
I don't know why, but there is already an existing function in this very same file, that is called by clicking on the "Show desktop" taskbar context menu item, whose handling is there:
https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/explorer/traywnd.cpp;hb=34635e1585e97539358d975c4fe0e03758c3c14d#l612
(and related: https://git.reactos.org/?p=reactos.git;a=blob;f=base/shell/explorer/traywnd.cpp;hb=34635e1585e97539358d975c4fe0e03758c3c14d#l568 )
Why not calling those functions directly instead?

Reply:
`CTrayShowDesktopButton` is a child window.
Now we use `WM_COMMAND:TRAYCMD_TOGGLE_DESKTOP` instead.

Reply:
In any case you could call those `ShowDesktop()` etc. functions?

Reply:
`PostMessage` function does add a message onto the message queue. Thus, it can be delayed the action when the system is too busy. It is good for the user to queue mouse input messages as expected behaviour.

Reply:
`ShowDesktop()` is in class `CTrayWindow`. This code is in class `CTrayShowDesktopButton`.

Reply:
How does it behave with double-clicks?

Reply:
Now, we have no `CS_DBLCLKS` flag. So double-clicks are converted to single clicks.